flexget/components/managed_lists/lists/ombi_list.py

```diff
@@ -0,0 +1,616 @@
+
+    def __init__(
+        self, base_url: str, auth: Callable[[], Dict[str, str]], data: Dict[str, Any]
+    ) -> None:
+        super().__init__(base_url, auth, self.entry_type, data)
+
+    @classmethod
+    def from_tvdb_id(cls, base_url: str, tmdb_id: str, auth: Callable[[], Dict[str, str]]):
+        """Create a Ombi Entry from an TVDB ID."""
+
+        headers = {'Content-Type': 'application/json', 'Accept': 'application/json'}
+        headers.update(auth())
+
+        url = f"{base_url}/api/v2/Search/{cls.entry_type}/moviedb/{tmdb_id}"
+
+        response = requests.get(url, headers=headers)
+
+        if not response.ok:
+            log.error(f"{url} return {response.reason}")
+            # Do I throw an exception here? Is that allowed from a plugin?
+            return None
+
+        data = response.json()
+
+        # Their is a bug in OMBI where if you get a movie by its TMDB ID
+        # then the theMovieDbId field will be blank for some reason...
+        data['theMovieDbId'] = tmdb_id
+
+        return OmbiTv(base_url, auth, data)
+
+
+class OmbiSet(MutableSet):
+    """The schema for the Ombi managed list."""
+
+    supported_ids = ['imdb_id', 'tmdb_id', 'ombi_id']
+    schema = {
+        'type': 'object',
+        'properties': {
+            'url': {'type': 'string'},
+            'api_key': {'type': 'string'},
+            'username': {'type': 'string'},
+            'password': {'type': 'string'},
+            'type': {'type': 'string', 'enum': ['shows', 'seasons', 'episodes', 'movies']},
+            'only_approved': {'type': 'boolean', 'default': True},
+            'include_available': {'type': 'boolean', 'default': False},
+            'include_year': {'type': 'boolean', 'default': False},
+            'include_ep_title': {'type': 'boolean', 'default': False},
+        },
+        'oneOf': [{'required': ['username', 'password']}, {'required': ['api_key']}],
+        'required': ['url', 'type'],
+        'additionalProperties': False,
+    }
+
+    @property
+    def immutable(self):
+        return False
+
+    def __init__(self, config: Dict[str, Any]):
+        self.config = config
+        self._items = None
+        self._cached_items = None
+
+    def __iter__(self):
+        return iter(self.items)
+
+    def __len__(self):
+        return len(self.items)
+
+    def add(self, entry: Entry):
+        log.info(f"Adding {entry['title']} to ombi_list.")
+
+        log.debug(f"Getting OMBI entry for {entry['title']}.")
+
+        ombi_entry = self._get_ombi_entry(entry)
+
+        if not ombi_entry:
+            log.error(f"Failed to find OMBI entry for {entry['title']}.")
+            return
+
+        ombi_entry.mark_requested()
+
+        log.info(f"{entry['title']} was added to ombi_list.")
+
+    def __ior__(self, entries: List[Entry]):
+        for entry in entries:
+            self.add(entry)
+
+    def discard(self, entry: Entry):
+        """Removes an entry from OMBI by marking it as available.
```

Comment:
Maybe we need to discuss expected behaviour or add a configuration item called remove_behaviour: with options such as mark_available, delete from list

I didn't see this at first and was very surprise that when I ran [list_remove](https://flexget.com/Plugins/List/list_remove) against my list everything became marked available instead of being removed from ombi.

I can see having both behaviours valuable.

Reply:
It's the same with marking something as requested. Once it's marked as requested by a user, do you want to approve it as well?

While I think we should eventually support all these features, for a first pass just to get it working and released maybe stick to a more simple approach first?

Reply:
> I am all for getting the minimum viable out but having a config item that shows what the possible options seems like a good idea.

But we should probably add the config option when we add the functionality and not before the feature exists. I do think it will eventually make sense to allow you to override what is happening when  you remove items from the list.

> instead it marked them unavailable which was unexpected..

It should not have but it could be a bug. Here is the code just a little further down. 

```python
ombi_entry.mark_requested()
ombi_entry.mark_available()

log.info(f"{entry['title']} was removed from ombi_list.")
```

> I can't directly work on your branch without permissions.

Fork my fork. https://github.com/shadycuz/Flexget/tree/ombi_managed_list then checkout `ombi_managed_list` and once you have it checked out locally you can make a new branch from it with `git checkout -b new_branch_name`. 

After you push up changes, GitHun should nag you in the GUI about opening a PR against my fork. Just make sure to point your PR at `ombi_managed_list`.

Reply:
I was thinking about making an overseerr plugin and wordering about similar issues regarding what status an added or removed entry should go to. I think one of the most common uses would be that the list contains all 'approved' items, adding puts it in the 'approved' status, and removing puts it into 'available'. There are certainly many other cases we'd want to be able to control though. I think probably the most flexible config schema would involve 3 options:
```yaml
# Define what statuses should show up as items in the list
statuses:
- unapproved
- approved
# What status a newly added entry should be
add_status: approved
# What status an entry 'removed' from the list would go to. We'd also need something like 'removed' to be an option to actually fully delete it from ombi/overseerr
remove_status: available
```
I don't really know if this is the most user friendly option though. Still mulling it over.

Reply:
> I think one of the most common uses would be that the list contains all 'approved' items

Thats interesting. I think the most common use case would be that the list contains all items that have been requested by users. Maybe overseer works differently than ombi. In ombi you can request something, but then it needs approved.  I think that the most common use case for ombi is to allow your close friends and family to request media to be added to your plex server that you shared with them. In this use case, approvals aren't really needed. The main purpose the ombi_list would serve is to see what people want and to let them know once its avaliable.

I do see the use case and in the future it might make sense to offer the functionality by making the list configurable. I just don't want to start down that rabbit hole yet or this PR will be over two years old by the time it gets merged.

but I do like you example config options =)

Reply:
Overseerr has options where requests go directly into the approved category for users you specify, (or everybody,) so I think the workflow is the same in that case since that list is identical to the all requested list. My main point is that in the common case for flexget, you'd want a list of items that users want, and removing them from the list would signify that they have been downloaded. I think most people using something like ombi or overseerr would then do most of the other managment tasks directly from the ombi/overseerr interface.

Reply:
> My main point is that in the common case for flexget, you'd want a list of items that users want, and removing them from the list would signify that they have been downloaded

Thats exactly how I feel and what I thought I just described? Also its how the ombi list currently behaviors right now.

Reply:
Yeah, we are in agreement then. I think if that's the default behavior then allowing all the other options are icing on the cake. Would let you do other management tasks from flexget, (other than the downloading part,) with a lot of configurability.

Thinking about it more, we probably only need 2 possible actions when removing from the list. Move to 'available', and actually remove. Moving to other statuses might be better accomplished by adding it to another list. Maybe we just need a 'status' option, and some sort of flag to signify whether remove means delete/available.

```yaml
status: approved  # This could accept a list or a single string. When adding to the list it would go into the first status specified here.
delete_on_remove: no  # Don't like this name, but this would control whether something got actually deleted or moved to 'available' status.
```
Some theoretical task snippets:
```yaml
# Typical use case. Match requests, mark available
list_match:
- ombi_list:
    status: [approved, requested] # If adding to the list, it would go into the status specified first here
    delete_on_remove: no #this would be default
```
```yaml
# auto approve all requests
ombi_list:
  status: requested
accept_all: yes
list_add:
- ombi_list:
    status: approved
```
(I might be slightly off with the statuses in ombi, I haven't used it.)

Reply:
I like these ideas and also have considered using overseer vs ombi a few times but didn't because I had my ombi plugin already working and didn't have time.

The functionality is basically the same between them other than small differences.

Also I agree that the "DEFAULT' would make sense to be "available" when an item is removed as YES that is what flexget does by default. The funny thing is that both Ombi and overseer scan PLEX so in my system that gets done automatically anyway, also one advantage of doing it in plex is it assumes your download is successful at that point and it may trigger an available email to an ombi user WAY before the file has completed download.

I don't want to expand scope but if it makes a difference here for future through.. Ombi supports music as well.. Flexget doesn't directly support music but a future enhancement I was thinking of was getting the music requests.

Reply:
> . The funny thing is that both Ombi and overseer scan PLEX so in my system that gets done automatically anyway, also one advantage of doing it in plex vs flexget is flexget assumes your download is successful at that point and it may trigger an available email to an ombi user WAY before the file has completed download.

Oh, I forgot about that. Overseerr actually solves that by having a 'processing' state which should probably be the default on remove for that (possible future) plugin.

Reply:
I haven't tried that on mine, but will look at it tomorrow if I get time.

The real challenge is going to be TV shows. Show, Season, Episode. Mainly because the OMBI API is really rough.

Reply:
Same goes for the earlier question about approval. BOTH Overseer and Ombi have automatic or manual approval options built in IIRC so it is unlike you need to duplicate a way to approve items in flexget. Nether one has a bulk way to delete or expire requests from what I recall other than if they are marked available. Hence why I would like both functions in the flexget list.

Reply:
@gazpachoking @bagobones 

Today I fixed the cache mechanism and refactored how requested items are filtered before they are added to `self._items`. I implemented a TypedDict for better type hinting of the plugins `self.config` attribute. I also added the `status` config field.

```
'status': {
                'type': 'string',
                'enum': ['approved', 'available', 'requested', 'denied'],
                'default': 'requested',
            },
```

This will control what items are added to the list from Ombi when the plugin is first initialized. I haven't refactored the `add` or `discard` methods yet. But I did realize today that we could reuse the `status` for when we are `adding` items to the list. For example, if I set `status` to `denied` then the `ombi_list` will be a list of `denied` items in Ombi. So if you want to `add` items from Flexget into `ombi_list` then obviously they would have to be added and set to `denied` as well? Let me know what you both think about that but it seems to make sense to me. 

Once that is complete then I will move on to the `remove_status`. I'm not even sure what all the options are. I know you can `deny` a request. I think you can also `delete` a request so it goes back to "unrequested" status which is the "default" or "normal" status for items in Ombi.

Also wanted to let you all know how I thankful I am for Flexget. I have been a user for like 5 to 8 years now? It has saved me so much time and effort <3. Glad I can finally give back, even if its taken a while to get this merged lol.

Reply:
@shadycuz at this rate you will have completed refactoring this plugin before I can contribute.. Maybe I should just stay  an advanced tester instead of fighting with my debug / git setup. Thank YOU for taking the ombi integration further..

>Also wanted to let you all know how I thankful I am for Flexget. I have been a user for like 5 to 8 years now? It has saved me so much time and effort <3. Glad I can finally give back, even if its taken a while to get this merged lol.

On that note I think it saved me from having a melt down with my significant other who was continually requesting if I could get things... the automation worked too well however. She managed to fill the NAS a few times lol.

Reply:
@bagobones  @gazpachoking 
Today I added the ability to add items from Flexget into Ombi with whatever status has been configured for the list. For example:
```yaml
list_add:
      - ombi_list:
          type: movies
          status: available
```
Would mark items as available in Ombi. One of the things I discovered today was transitioning from one status to another is not a straightforward path. This is because an entry in Ombi can either be requested or not requested. The other statuses like `approved` or `denied` are a sub-status. The rules for how to transition are not clear but it appears you can go from `requested` to any sub status directly. So the following would work:
```yaml
# auto approve all requests
ombi_list:
  status: requested
accept_all: yes
list_add:
- ombi_list:
    status: approved
```

but you likely would not be able to go from denied to approved/available. At least not directly without other logic. This probably won't work in the plugin's current state:
```yaml
# auto approve all requests
ombi_list:
  status: denied
accept_all: yes
list_add:
- ombi_list:
    status: approved
```

I will soon start on the ability to delete an item from Ombi, which would mean resetting its status back to `unrequested` but that might be a bit complicated depending on the item's current status.

Reply:
@bagobones  @gazpachoking 

I have implemented the ability to remove items using specific status by setting `on_remove` to one of the following: `unavailable', 'denied', 'deleted'`. The default is to delete which basically removes the `request` from Ombi which makes it a "normal" item. 

I did learn a lot. For example an Item in Ombi is either requested or not. If it is requested then it can have a sub status of `approved, denied, available`. This is interesting because some of these you can have at the exact same time. For example something can be `denied` while also being `available`. Basically the sub status comes in pairs. approved/denied, available/unavailable. You can also mix the pairs. 

I also found out that if you Sync your Plex with Ombi then those items behave differently than other items. For example if you have the movie "Top Gun" and it was synced from Plex and it shows as "available". It will show up in `ombi_plugin` but you can't discard it. So trying to mark it as `unavailable, deleted, denied` won't do anything.

I'm going to mark this conversation as resolved for now. Feel free to un-resolve it or add new review comments.